src/jobs/keys-update/staking-module-updater.service.ts

```diff
@@ -0,0 +1,194 @@
+          currNonce,
+          currentBlockHash,
+        );
+        continue;
+      }
+
+      if (await this.isReorgDetected(updaterState, prevBlockHash, currentBlockHash)) {
+        this.logger.log('Reorg detected, start indexing', { stakingModuleAddress, currentBlockHash });
+
+        await this.updateStakingModule(
+          updaterState,
+          moduleInstance,
+          contractModule,
+          stakingModuleAddress,
+          currNonce,
+          currentBlockHash,
+        );
+        continue;
+      }
+
+      if (
+        prevElMeta.blockNumber < currElMeta.number &&
+        (await moduleInstance.operatorsWereChanged(
+          contractModule.stakingModuleAddress,
+          prevElMeta.blockNumber + 1,
+          currElMeta.number,
+        ))
+      ) {
+        this.logger.log('Update operator events happened, need to update operators', {
+          stakingModuleAddress,
+          currentBlockHash,
+        });
+
+        await this.updateStakingModule(
+          updaterState,
+          moduleInstance,
+          contractModule,
+          stakingModuleAddress,
+          currNonce,
+          currentBlockHash,
+        );
+        continue;
+      }
+
+      this.logger.log('No changes have been detected in the module, indexing is not required', {
+        stakingModuleAddress,
+        currentBlockHash,
+      });
+    }
+
+    // Update EL meta in db
+    await this.elMetaStorage.update({ ...currElMeta, lastChangedBlockHash: updaterState.lastChangedBlockHash });
+  }
+
+  public async updateStakingModule(
+    updaterState: UpdaterState,
+    moduleInstance: StakingModuleInterface,
+    contractModule: StakingModule,
+    stakingModuleAddress: string,
+    currNonce: number,
+    currentBlockHash: string,
+  ) {
+    await moduleInstance.update(stakingModuleAddress, currentBlockHash);
+    await this.srModulesStorage.upsert(contractModule, currNonce, currentBlockHash);
+    updaterState.lastChangedBlockHash = currentBlockHash;
+  }
+
+  public async isReorgDetected(updaterState: UpdaterState, prevBlockHash: string, currentBlockHash: string) {
+    // calculate once per iteration
+    // no need to recheck each module separately
+    if (updaterState.isReorgDetected) {
+      return true;
+    }
+
+    const currentBlock = await this.executionProvider.getFullBlock(currentBlockHash);
+    const prevBlock = await this.executionProvider.getFullBlock(prevBlockHash);
+
+    if (currentBlock.parentHash === prevBlock.hash) return false;
+    // different hash but same number
+    if (currentBlock.hash !== prevBlock.hash && currentBlock.number === prevBlock.number) {
+      updaterState.isReorgDetected = true;
+      return true;
+    }
+
+    const blocks = await Promise.all(
+      range(prevBlock.number, currentBlock.number + 1).map(async (bNumber) => {
+        return await this.executionProvider.getFullBlock(bNumber);
+      }),
+    );
+
+    for (let i = 1; i < blocks.length; i++) {
```

Comment:
i think we lost here check before loop: prevBlockHash !== block[0].hash, discussed it on call

Reply:
If prevBlockHash !== blocks[0].hash evaluates to true, this indicates a reorganization has been found. This is because we have blocks with the same blockNumber but differing hashes;

However, if prevBlockHash !== blocks[0].hash evaluates to false (meaning the hashes are the same), it does not  mean that a reorganization has not occurred

Reply:
I've added these checks

Reply:
What do you think about adding comment to every check ? i think it could help to check algorithm and even find mistake in logic
i mean `blocks[0].hash !== prevBlockHash` - means we found reorg

but here i think i need more deep understanding
``` 
if (currentBlock.parentHash !== previousBlock.hash) {
     updaterState.isReorgDetected = true;
     return true;
}
```

Reply:
added